base/abstractarray.jl

```diff
@@ -2057,124 +2057,51 @@ push!(A, a, b, c...) = push!(push!(A, a, b), c...)
-        n = 1
-        local step, laststep, laststate
-
-        h = hash(x1, h)
-        h += hashr_seed
-
-        while true
-            # If overflow happens with entries of the same type, a cannot be equal
-            # to a range with more than two elements because more extreme values
-            # cannot be represented. We must still hash the two first values as a
-            # range since they can always be considered as such (in a wider type)
-            if isconcretetype(T)
-                try
-                    step = x2 - x1
-                catch err
-                    isa(err, OverflowError) || rethrow(err)
-                    break
-                end
-                # If true, wraparound overflow happened
-                sign(step) == cmp(x2, x1) || break
-            else
-                applicable(-, x2, x1) || break
-                # widen() is here to ensure no overflow can happen
-                step = widen(x2) - widen(x1)
-            end
-            n > 1 && !isequal(step, laststep) && break
-            n += 1
-            laststep = step
-            if y === nothing
-                # The array matches a range exactly
-                return hash(x2, hash(n, h))
-            end
-            x1, x2 = x2, y[1]
-            y = iterate(a, y[2])
-        end
-
-        # Always hash at least the two first elements as a range (even in case of overflow)
-        if n < 2
-            h = hash(2, h)
-            h = hash(y2[1], h)
-            @assert y !== nothing
-            x1, x2 = x2, y[1]
-            y = iterate(a, y[2])
-        else
-            h = hash(n, h)
-            h = hash(x1, h)
-        end
+## hashing AbstractArray ##
+
+function hash(A::AbstractArray, h::UInt)
+    h = hash(AbstractArray, h)
+    # Axes are themselves AbstractArrays, so hashing them directly would stack overflow
+    # Instead hash the tuple of firsts and lasts along each dimension
+    h = hash(map(first, axes(A)), h)
+    h = hash(map(last, axes(A)), h)
+    isempty(A) && return h
+
+    # Now work backwards and hash (up to) three distinct key-value pairs
+    # Working backwards introduces an asymmetry with isequal; in many cases
+    # arrays that hash equally will be compared via isequal, which iteratively
+    # works forwards and _short-circuits_. Therefore the elements at the
+    # beginning of the array are not as valuable to include in the hash computation
+    # as they are "cheaper" to compare within `isequal`.
+    # A small number of distinct elements are included in the hashing algorithm
+    # in order to emphasize distinctions between arrays that are nearly all the
+    # same constant value but have a handful of differences the O(log(n)) skipping
+    # algorithm might miss (in particular, this includes sparse matrices).
+    I = keys(A)
+    i = last(I)
+    v1 = A[i]
+    h = hash(i=>v1, h)
+    i = let v1=v1; findprev(x->!isequal(x, v1), A, i); end
```

Comment:
possibly should use `Ref{eltype(A)}(v1)` here, or just open-code the loop (to make heterogeneous arrays fast). the same would apply to `h = hash(i=>v1, h)` though (vs. writing it as `h = hash(i, hash(v1, h))`), so I'm not sure of the importance of it

Reply:
I added a simple type-unstable case to the matrix above — this is already smoking the status quo (largely by simply touching fewer unstable elements). I get your point about constructing the `Pair`s, but I don't fully understand the `Ref` in the closure.  In any case, this should be a relatively small constant cost.